modules/html/src/main.rs

```diff
@@ -93,32 +105,64 @@ fn transform_heading(heading: Value) -> String {
         }
     }
 
-    write!(result, r#"{{"name": "raw", "data": "</h{level}>"}}"#,).unwrap();
+    write!(result, "{},", raw!(format!("</h{level}>"))).unwrap();
+    result.push(']');
+
+    result
+}
+
+fn transform_error(error: Value) -> String {
+    let mut result = String::new();
+    result.push('[');
+
+    let Value::String(source) = &error["arguments"]["source"] else {
+        panic!();
+    };
+    let Value::String(input) = &error["arguments"]["input"] else {
+        panic!();
+    };
+    let Value::String(err) = &error["data"] else {
+        panic!();
+    };
+
+    // TODO: Maybe make these errors look better. Be careful though, see notes in API, don't use
+    //   calls to other modules that may fail. I have taken care to not use __text but rather just
+    //   entered the text myself, because if I used __text and that failed, it would lead to
+    //   infinite recursion, which is bad
+    write!(result, "{},", raw!(r#"<span style="background:#FF0000">"#)).unwrap();
+
+    let data = escape(format!("Error originating from {source}: {err} on input {input}").as_str());
+    write!(result, "{},", raw!(data)).unwrap();
+
+    write!(result, "{}", raw!("</span>")).unwrap();
+
     result.push(']');
 
     result
 }
 
 fn escape_text(module: Value) -> String {
     if let Value::String(s) = &module["data"] {
-        let s = s
-            .replace('&', "&amp;")
-            .replace('<', "&lt;")
-            .replace('>', "&gt;")
-            .replace('"', "&quot;")
-            .replace('\'', "&#39;")
-            .replace("\r\n", "\\n")
-            .replace('\n', "\\n");
-        format!(r#"[{{"name": "raw", "data": "{s}"}}]"#)
+        let s = escape(s);
+        format!("[{}]", raw!(s).to_string())
     } else {
         panic!("Malformed text module");
     }
 }
 
+fn escape(text: &str) -> String {
```

Comment:
maybe a char instead of a &str as parameter

Reply:
Hmm, I think this is the most common use case, but I can do a separate function which just escapes a character, which this function delegates to. I doubt it will be very useful, and I think this may be more efficient since it is just one string and not multiple strings that has to be combined, but if you think there will be a need for a function taking a char, I will do it!